src/Bridge/Doctrine/MongoDB/Extension/PaginationExtension.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ApiPlatform\Core\Bridge\Doctrine\MongoDB\Extension;
+
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+use Doctrine\Common\Persistence\ManagerRegistry;
+use Doctrine\ODM\MongoDB\Query\Builder;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use ApiPlatform\Core\Bridge\Doctrine\MongoDB\Paginator;
+
+/**
+ * Applies pagination on the Doctrine query for resource collection when enabled.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ * @author Samuel ROZE <samuel.roze@gmail.com>
+ */
+class PaginationExtension implements QueryResultExtensionInterface
+{
+    private $managerRegistry;
+    private $requestStack;
+    private $resourceMetadataFactory;
+    private $enabled;
+    private $clientEnabled;
+    private $clientItemsPerPage;
+    private $itemsPerPage;
+    private $pageParameterName;
+    private $enabledParameterName;
+    private $itemsPerPageParameterName;
+
+    public function __construct(ManagerRegistry $managerRegistry, RequestStack $requestStack, ResourceMetadataFactoryInterface $resourceMetadataFactory, bool $enabled = true, bool $clientEnabled = false, bool $clientItemsPerPage = false, int $itemsPerPage = 30, string $pageParameterName = 'page', string $enabledParameterName = 'pagination', string $itemsPerPageParameterName = 'itemsPerPage')
+    {
+        $this->managerRegistry = $managerRegistry;
+        $this->requestStack = $requestStack;
+        $this->resourceMetadataFactory = $resourceMetadataFactory;
+        $this->enabled = $enabled;
+        $this->clientEnabled = $clientEnabled;
+        $this->clientItemsPerPage = $clientItemsPerPage;
+        $this->itemsPerPage = $itemsPerPage;
```

Comment:
To me, we have too many parameters here. That would be nice to create a normalised `PaginationConfiguration` object, that we could use both here and in the Doctrine implementation.